tunedasd/man/tunedasd.8
.\" Copyright 2017 IBM Corp.
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.TH TUNEDASD 8 "Apr 2006" "s390-tools"
.SH NAME
tunedasd \- A tool to adjust tunable parameters on DASD devices

.SH SYNOPSIS
.B tunedasd
[OPTION] COMMAND device

.SH DESCRIPTION
.B tunedasd
is a tool to adjust tunable parameters on DASD devices.
It can be used to modify the status of the device and to adjust
some performance relevant settings.
The DEVICE is the node of the device (e.g. '/dev/dasda') or a list of
devices separated by a space character.
Following functions are supported:
.IP "     -"
Get and reset the performance statistics profile per device.
.IP "     -"
Get and set the caching mode of the storage server.
This is only available for ECKD devices, because it is implemented using
ECKD Channel Program commands.
.IP "     -"
Reserve/Release/Unconditional Reserve a device.
This is only available for ECKD devices, because it is based on
the Control Unit functions.
.P
If
.B tunedasd
is called without parameter, it prints a short usage information and exits.
.SH OPTIONS
.TP
.BR "\-v" " or " "\-\-version"
Print version information, then exit.
.TP
.SH COMMANDS
.TP
.BR "\-h" " or " "\-\-help"
Print usage information, then exit.
.TP
.BR "\-g" " or " "\-\-get_cache"
Get current storage server caching behavior.
.TP
.BR "\-c" " or " "\-\-cache" " <behavior> "
Define caching behavior on storage server.
The following caching behaviors are currently supported by modern
Enterprise Storage Servers (ESS):
.br
.IP "          \(bu" 12
.I normal
('Normal Cache replacement'),
.IP "          \(bu" 12
.I bypass
(Bypass Cache),
.IP "          \(bu" 12
.I inhibit
(Inhibit Cache Loading),
.IP "          \(bu" 12
.I sequential
(Sequential Access),
.IP "          \(bu" 12
.I prestage
(Sequential Prestage),
.IP "          \(bu" 12
.I record
(Record Access)
.IP "" 7
More details about caching can be found in the 'Storage Control Reference' of the attached storage server.
.TP
.BR "\-n" " or " "\-\-no_cyl" " <n> "
Number of cylinders to be cached (only valid together with --cache).
.TP
.BR "\-S" " or " "\-\-reserve"
Reserve device.
.TP
.BR "\-L" " or " "\-\-release"
Release device.
.TP
.BR "\-O" " or " "\-\-slock"
Unconditional reserve device.
.br
.B Note:
Use with care, this breaks an existing reserve.
.TP
.BR "\-\-enable-stats"
Enable performance statistics globally.
.br
\fBNote:\fR Same as "echo 'set on' > /proc/dasd/statistics"
.TP
.BR "\-\-disable-stats"
Disable performance statistics globally.
.br
\fBNote:\fR Same as "echo 'set off' > /proc/dasd/statistics"
.TP
.BR "\-P" " or " "\-\-profile"
Print profile info of device.
.br
.B Note:
Profile info must be available and enabled in the kernel to get valid results
out of the profile commands. Enable statistics via \fB\-\-enable\-stats\fR.
.TP
.BR "\-Q" " or " "\-\-query_reserve"
Query the current reserve status of the device.
The following states are defined:
.br
.IP "          \(bu" 12
.I none:
The device is not reserved.
.IP "          \(bu" 12
.I implicit:
The device is not reserved, but there is a contingent or implicit
allegiance to this Linux instance.
.IP "          \(bu" 12
.I other:
The device is reserved to another operating system instance.
.IP "          \(bu" 12
.I reserved:
The device is reserved to this Linux instance.
.IP "" 7
More details about reserve/release can be found in the 'Storage Control Reference' of the attached storage server.

.TP
.BR "\-I" " or " "\-\-prof_item <row> "
Print single profile item data row (without header).
.br
Following rows are supported:
.br
.IP "          \(bu" 12
.I reqs
(# of dasd I/O requests),
.IP "          \(bu" 12
.I sects
(# of 512Byte sectors),
.IP "          \(bu" 12
.I sizes
(Histogram of sizes),
.IP "          \(bu" 12
.I total
(Histogram of I/O times),
.IP "          \(bu" 12
.I totsect
(Histogram of I/O times per sector),
.IP "          \(bu" 12
.I start
(Histogram of I/O time till ssch),
.IP "          \(bu" 12
.I irq
(istogram of I/O time between ssch and irq),
.IP "          \(bu" 12
.I irqsect
(istogram of I/O time between ssch and irq per sector),
.IP "          \(bu" 12
.I end
(Histogram of I/O time between irq and end),
.IP "          \(bu" 12
.I queue
(# of req in chanq at enqueuing)
.TP
.BR "\-R" " or " "\-\-reset_prof"
Reset profile info of device.
.TP
.BR "\-p" " or " "\-\-path_reset <chpid>"
Reset a channel path <chpid> of a selected device. A channel path
might be suspended due to high IFCC error rates or a High Performance
FICON failure. Use this option to resume considering the channel path
for I/O.
.TP
.BR "\-\-path_reset_all"
Reset all channel paths of the selected device. The channel paths
might be suspended due to high IFCC error rates or a High Performance
FICON failure. Use this option to resume considering all defined
channel paths for I/O.
.TP
.BR "\-s" " or " "\-\-copy\-pair\-swap <copy_pair>"
Swap the copy roles of the specified copy pair. The <copy_pair> has to
be a comma separated pair of \fBPRIMARY,SECONDARY\fR where
\fBPRIMARY\fR is the old primary device that will become a secondary
automatically. The old \fBSECONDARY\fR device will become the new
primary device. Both devices have to be online for this operation to
succeed.
.\"
.\".TP
.\".BR "\-o" " or " "\-\-online"
.\"Set the device online using the sysfs interface.

.SH EXAMPLE
1. Scenario: Get profile info for a device.
.br

       tunedasd \-P /dev/dasdc
       tunedasd \-PI irq /dev/dasdc

.br
2. Scenario: Set device caching mode to 1 cylinder 'prestage'.
.br

       tunedasd \-c prestage \-n 1 /dev/dasdc

.br
3. Scenario: Reset failed channel path with CHPID 45
.br

       tunedasd \-p 45 /dev/dasdc

.br
4. Scenario: Swap copy pair 0.0.9700 and 0.0.9740
.br

       tunedasd \-s 0.0.9700,0.0.9740 /dev/dasdc
.br
.SH "SEE ALSO"
.BR dasdview (8),
.BR dasdfmt (8),
.BR fdasd (8)


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.xx.x (2023-xx-xx)__

  For Linux kernel version: 6.x

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.32.0 (2024-04-03)__

  For Linux kernel version: 6.8

  Changes of existing tools:
  - cpumf/lscpumf: add support for machine type 3932
  - genprotimg, pvattest, and pvsecret accept IBM signing key with Armonk as
    subject locality
  - zdump/zipl: Support for List-Directed dump from ECKD DASD
  - zkey: Detect FIPS mode and generate PBKDF for luksFormat according to it

  Bug Fixes:
  - dbginfo.sh: dash compatible copy sequence
  - rust/pv_core: Fix UvDeviceInfo::get() method
  - zipl/src: Fix leak of files if run with a broken configuration
  - zkey: Fix convert command to accept only keys of type CCA-AESDATA

* __v2.31.0 (2024-02-02)__

  For Linux kernel version: 6.7

  General:
  - common.mak: Set default C/C++ standard to gnu11/gnu++11

  Add new tools / libraries:
  - pvapconfig: Tool to automatically configure APQNs in SE KVM guests
  - s390-tools: Provide pre-commit configuration

  Changes of existing tools:
  - cpuplugd: Adjust to CPU 0 being no longer hotpluggable
  - dbginfo.sh: Check for Dynamic Partition Mode
  - dbginfo.sh: Update man page and copyright
  - rust/pv: Add user-data signing and verifying
  - rust/pvsecret: Add user defined signatures and verifications
  - zdev/dracut: Consolidate device configuration

  Bug Fixes:
  - dbginfo.sh: Fix relative path on script copy
  - libkmipclient: Fix build with libxml2-2.12.0
  - pvsecret: Fix panic if empty file is used as host key document
  - rust/pv: Fix 'elided_lifetimes_in_associated_constant' warning

* __v2.30.0 (2023-12-01)__

  For Linux kernel version: 6.6

  Add new tools / libraries:
  - lspai: Tool to display PAI counter sets
  - s390-tools: Provide a ShellCheck configuration

  Changes of existing tools / libraries:
  - cpumf/pai: Add command line option for realtime scheduling
  - dbginfo.sh: enhance ethtool collection for ROCE
  - libutil/util_lockfile: add routine to return owning pid of file lock
  - lszcrypt: Improve lszcrypt output on SE guests
  - rust: Use a single workspace for all rust tools
  - zdev: limit the derivation of ZDEV_SITE_ID
  - zdump/df_s390: Update 'zgetdump -i' output with zlib info
  - zdump/dfi_s390: Support reading compressed s390_ext dumps
  - zipl/boot: Integrate zlib compression to single volume DASD dumper
  - zipl/boot: compile the bootloaders only if HOST_ARCH is s390x
  - zipl: Add --no-compress option to zipl command
  - zkey: Also check for deconfigured and check-stopped cards
  - dbginfo.sh: fix relative path on script copy

  Bug Fixes:
  - ap_tools/ap-check: handle get-attributes between pre and post event
  - libutil: fix util_file_read_*() using wrong format specifiers
  - rust/pv: fix Invalid write of size 1

* __v2.29.0 (2023-08-04)__

  For Linux kernel version: 6.5

  General:
  - s390-tools now supports tools written in Rust.
  - Add `compdb` Makefile target to create 'compile_commands.json' to LSP
    backends in IDEs and editors

  Add new tools / libraries:
  - rust/pv: Library for pv tools written in rust
  - rust/pvsecret: Tool to manage UV-secrets

  Changes of existing tools:
  - dbginfo.sh: Global IFS variable
  - genprotimg: Add support for add-secret requests
  - genprotimg: Build debuginfo files for bootloader
  - hyptop: Add real SMT utilization field
  - hyptop: Allow users to set speedup factor
  - pvattest: Add yaml-output for verify command
  - zipl: Build debuginfo files for bootloader

  Bug Fixes:
  - dump2tar: Fix truncated paths
  - zdev/dracut: fix kdump build to integrate with site support

* __v2.28.0 (2023-07-11)__

  For Linux kernel version: 6.4

  Changes of existing tools:
  - chzcrypt: Support for SE AP pass-through support
  - genprotimg: Add support for non-s390x architectures
  - lszcrypt: Support for SE AP pass-through support
  - zdev: Add support for autoquiesce related sysfs attributes

  Bug Fixes:
  - ap_tools/ap-check: Handle missing 'matrix' and 'control_domains' attrs
  - ap_tools/ap-check: Hold ap config file lock over get attributes
  - s390-tools: Fix build for ppc64le
  - zdev: Add missing label in the udev-rules
  - zdev: Add proper value input for the ZDEV_SITE_ID key
  - zdev: Use rename-file to avoid any symlinks created
  - zipl/dump: fix ngdump dracut helper script

* __v2.27.0 (2023-05-30)__

  For Linux kernel version: 6.3

  Changes of existing tools:
   - s390-tools cross-compile and non-s390x support:
      - `pkg-config` is now mandatory for the build process
      - Add `PKG_CONFIG` Makefile variable to select pkg-config program;
        default `pkg-config` or `$(CROSS_COMPILE)pkg-config` if
        `CROSS_COMPILE` is set
      - Rename Makefile variable `ARCH` to `HOST_ARCH`. `HOST_ARCH` is the
        architecture that will run the produced (executable) objects
      - Add the Makefile variable `BUILD_ARCH`. `BUILD_ARCH` is the
        architecture of the build system. For each Makefile variable like
        `CC`, `LINK`, `CPP`, ... there is a suffixed version of it - e.g.
        `CC_FOR_BUILD`. This is useful for cross compiling, and this naming
        convention is very similar to the Meson convention (see
        https://mesonbuild.com/Reference-tables.html#environment-variables-per-machine).
      - Limit build targets for non-s390x architectures (pvattest)
  - dasdfmt: Fall back to full format if space release fails
  - dbginfo.sh: Add nstat for network and SNMP stats
  - dbginfo.sh: Rework crypto data collection
  - hyptop: Show thread util by default
  - zipl: Add support for list-directed IPL dump from ECKD DASD

  Bug Fixes:
  - lszcrypt: Fix argument parsing
  - zdev/dracut: Fix out-of-memory (OOM) situations in the kdump crashkernel environment
  - ziomon/ziorep_config: Fix for SCSI devices of type disk without block dev

tunedasd/src/tunedasd.c
/*
 * tunedasd - Adjust tunable parameters on DASD devices
 *
 * Copyright IBM Corp. 2004, 2017
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <errno.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "lib/util_file.h"
#include "lib/util_opt.h"
#include "lib/util_prg.h"
#include "lib/zt_common.h"

#include "disk.h"
#include "tunedasd.h"

static const struct util_prg prg = {
	.desc = "Adjust tunable DASD parameters. More than one DEVICE node can "
		"be specified, but at least one (e.g. /dev/dasda)",
	.args = "DEVICE1 [DEVICE2...]",
	.copyright_vec = {
		{
			.owner = "IBM Corp.",
			.pub_first = 1999,
			.pub_last = 2017,
		},
		UTIL_PRG_COPYRIGHT_END
	}
};

/* Defines for options with no short command */
#define OPT_PATH_RESET_ALL	128
#define OPT_ENABLE_STATS	129
#define OPT_DISABLE_STATS	130

static struct util_opt opt_vec[] = {
	UTIL_OPT_SECTION("CACHING MODES (ECKD ONLY)"),
	{
		.option = { "cache", required_argument, NULL, 'c' },
		.argument = "BEHAVIOUR",
		.desc = "Specify caching behaviour on storage server: "
			"normal, bypass, inhibit, sequential, prestage, or record",
	},
	{
		.option = { "no_cyl", required_argument, NULL, 'n' },
		.argument = "NUM",
		.desc = "NUM cylinders to be cached (only valid with -c/--cache)",
	},
	{
		.option = { "get_cache", no_argument, NULL, 'g' },
		.desc = "Get current storage server caching behaviour",
	},
	UTIL_OPT_SECTION("RESERVE / RELEASE"),
	{
		.option = { "release", no_argument, NULL, 'L' },
		.desc = "Release device",
	},
	{
		.option = { "slock", no_argument, NULL, 'O' },
		.desc = "Unconditional reservce device\n"
			"NOTE: Use with care, this breaks an existing lock",
	},
	{
		.option = { "query_reserve", no_argument, NULL, 'Q' },
		.desc = "Print reserve status of device",
	},
	{
		.option = { "reserve", no_argument, NULL, 'S' },
		.desc = "Reserve device",
	},
	UTIL_OPT_SECTION("PERFORMANCE STATISTICS"),
	{
		.option = {
			"enable-stats", no_argument, NULL, OPT_ENABLE_STATS
		},
		.desc = "Enable performance statistics globally",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = {
			"disable-stats", no_argument, NULL, OPT_DISABLE_STATS
		},
		.desc = "Disable performance statistics globally",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = { "prof_item", required_argument, NULL, 'I' },
		.argument = "ROW",
		.desc = "Print single profile item: reqs, sects, sizes, total, "
			"totsect, start, irq, irqsect, end, or queue",
	},
	{
		.option = { "profile", no_argument, NULL, 'P' },
		.desc = "Print profile info of device",
	},
	{
		.option = { "reset_prof", no_argument, NULL, 'R' },
		.desc = "Reset profile info of device",
	},
	UTIL_OPT_SECTION("MISC"),
	{
		.option = { "path_reset", required_argument, NULL, 'p' },
		.argument = "CHPID",
		.desc = "Reset channel path CHPID of a device",
	},
	{
		.option = {
			"path_reset_all", no_argument, NULL, OPT_PATH_RESET_ALL
		},
		.desc = "Reset all channel paths of a device",
		.flags = UTIL_OPT_FLAG_NOSHORT,
	},
	{
		.option = { "copy-pair-swap", required_argument, NULL, 's' },
		.argument = "COPY_PAIR",
		.desc = "Swap a specified, comma separated copy pair.",
	},
	UTIL_OPT_HELP,
	UTIL_OPT_VERSION,
	UTIL_OPT_END
};

#define CMD_KEYWORD_NUM		17
#define DEVICES_NUM		256

enum cmd_keyword_id {
	cmd_keyword_help,
	cmd_keyword_version,
	cmd_keyword_get_cache,
	cmd_keyword_cache,
	cmd_keyword_no_cyl,
	cmd_keyword_reserve,
	cmd_keyword_release,
	cmd_keyword_slock,
	cmd_keyword_profile,
	cmd_keyword_prof_item,
	cmd_keyword_reset_prof,
	cmd_keyword_query_reserve,
	cmd_keyword_path,
	cmd_keyword_path_all,
	cmd_keyword_enable_stats,
	cmd_keyword_disable_stats,
	cmd_keyword_copy_swap,
};


/* Mapping of keyword IDs to strings */
static const struct {
	char* keyword;
	enum cmd_keyword_id id;
} keyword_list[] = {
	{ "help",           cmd_keyword_help },
	{ "version",        cmd_keyword_version },
	{ "get_cache",      cmd_keyword_get_cache },
	{ "cache",          cmd_keyword_cache },
	{ "no_cyl",         cmd_keyword_no_cyl },
	{ "reserve",        cmd_keyword_reserve },
	{ "release",        cmd_keyword_release },
	{ "slock",          cmd_keyword_slock },
	{ "profile",        cmd_keyword_profile },
	{ "prof_item",      cmd_keyword_prof_item },
	{ "reset_prof",     cmd_keyword_reset_prof },
	{ "query_reserve",  cmd_keyword_query_reserve },
	{ "path_reset",     cmd_keyword_path },
	{ "path_reset_all", cmd_keyword_path_all },
	{ "enable-stats",   cmd_keyword_enable_stats },
	{ "disable-stats",  cmd_keyword_disable_stats },
	{ "copy-swap",      cmd_keyword_copy_swap },
};	


enum cmd_key_state {
	req, /* Keyword is required */
	opt, /* Keyword is optional */
	inv  /* Keyword is invalid */
};


/* Determines which combination of keywords are valid */
static enum cmd_key_state cmd_key_table[CMD_KEYWORD_NUM][CMD_KEYWORD_NUM] = {
	/*		      help vers get_ cach no_c rese rele sloc prof prof rese quer path path enab disa copy
	 *		           ion  cach e    yl   rve  ase  k    ile  _ite t_pr y_re      _all le-s ble- -swa
	 *		               	e                                  m    of  serv            tats stat p
	 */
	/* help  	 */ { req, opt, opt, opt, opt, opt, opt, opt, opt, opt, opt, inv, inv, inv, inv, inv, inv },
	/* version	 */ { inv, req, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv },
	/* get_cache	 */ { opt, opt, req, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv },
	/* cache 	 */ { opt, opt, inv, req, opt, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv },
	/* no_cyl	 */ { opt, opt, inv, req, req, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv },
	/* reserve	 */ { opt, opt, inv, inv, inv, req, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv },
	/* release	 */ { opt, opt, inv, inv, inv, inv, req, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv },
	/* slock 	 */ { opt, opt, inv, inv, inv, inv, inv, req, inv, inv, inv, inv, inv, inv, inv, inv, inv },
	/* profile	 */ { opt, opt, inv, inv, inv, inv, inv, inv, req, opt, inv, inv, inv, inv, inv, inv, inv },
	/* prof_item	 */ { opt, opt, inv, inv, inv, inv, inv, inv, req, req, inv, inv, inv, inv, inv, inv, inv },
	/* reset_prof	 */ { opt, opt, inv, inv, inv, inv, inv, inv, inv, inv, req, inv, inv, inv, inv, inv, inv },
	/* query_reserve */ { inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, req, inv, inv, inv, inv, inv },
	/* path          */ { inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, req, inv, inv, inv, inv },
	/* path_all      */ { inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, req, inv, inv, inv },
	/* enable-stats  */ { inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, req, inv, inv },
	/* disable-stats */ { inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, req, inv },
	/* copy-swap     */ { inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, inv, req },
};

struct parameter {
	int kw_given;
	char *data;
};

struct command_line {
	struct parameter parm[CMD_KEYWORD_NUM];
	char * devices[DEVICES_NUM];
	int device_id;
};

/* Error message string */
#define ERROR_STRING_SIZE	1024
static char error_string[ERROR_STRING_SIZE];


/*
 * Generate and print an error message based on the formatted
 * text string FMT and a variable amount of extra arguments. 
 */
void
error_print (const char* fmt, ...)
{
	va_list args;

	va_start (args, fmt);
	vsnprintf (error_string, ERROR_STRING_SIZE, fmt, args);
	va_end (args);

	fprintf (stderr, "Error: %s\n", error_string);
}


/*
 * Print usage information.
 */
static void print_usage(void)
{
	util_prg_print_help();
	util_opt_print_help();
}

/* 
 * Check whether calling user is root. Return 0 if user is root, non-zero
 * otherwise. 
 */
static int
check_for_root (void)
{
	if (geteuid () != 0) {
		error_print ("Must be root to perform this operation");
		return -1;
	} else {
		return 0;
	}
}


/* 
 * Retrieve name of keyword identified by ID.
 */
static char *get_keyword_name(enum cmd_keyword_id id)
{
	unsigned int i;

	for (i = 0; i < sizeof (keyword_list) / sizeof (keyword_list[0]);
	     i++) {
		if (id == keyword_list[i].id) {
			return keyword_list[i].keyword;
		}
	}
	return "<unknown>";
}


/* 
 * Check the given function for given options and valid combinations of 
 * options
 */
static int check_key_state(struct command_line *cmdline)
{
	int i,j;

	/* Find first given keyword */
	for (i = 0; i < CMD_KEYWORD_NUM && !cmdline->parm[i].kw_given; i++);
	
	if (i >= CMD_KEYWORD_NUM) {
		error_print ("No valid parameter specified");
		print_usage ();
		return -1;
	}

	/* Check keywords */
	for (j = 0; j < CMD_KEYWORD_NUM; j++) {

		switch (cmd_key_table[i][j]) {
		case req:
			/* Missing keyword on command line */
			if (!(cmdline->parm[j].kw_given)) {
				error_print ("Option '%s' required when "
					     "specifying '%s'",
					     get_keyword_name (j),
					     get_keyword_name (i));
				return -1;
			}
			break;
		case inv:
			/* Invalid keyword on command line */
			if (cmdline->parm[j].kw_given) {
				error_print ("Only one of options '%s' and "
					     "'%s' allowed",
					     get_keyword_name (i),
					     get_keyword_name (j));
				return -1;
			}
			break;
		case opt:
			break;
		}
	}

	return 0;
}


/*
 * Save the given command together with its parameter. 
 */
static int
store_option (struct command_line* cmdline, enum cmd_keyword_id keyword,
	      char* value)
{
	if ((cmdline->parm[(int) keyword]).kw_given) {
		error_print ("Option '%s' specified more than once",
			     get_keyword_name (keyword));
		return -1;
	}
	cmdline->parm[(int) keyword].kw_given = 1;
	cmdline->parm[(int) keyword].data = value;
	return 0;
}


/*
 * Parse the command line for valid parameters.
 */
static int get_command_line(int argc, char *argv[], struct command_line *line)
{
	struct command_line cmdline;
	int opt;
	int rc;

	util_prg_init(&prg);
	util_opt_init(opt_vec, NULL);

	memset ((void *) &cmdline, 0, sizeof (struct command_line));

	/* Process options */
	do {
		opt = util_opt_getopt_long(argc, argv);

		rc = 0;
		switch (opt) {
		case 'h':
			rc = store_option (&cmdline, cmd_keyword_help,
					   optarg);
			break;
		case 'v':
			rc = store_option (&cmdline, cmd_keyword_version,
					   optarg);
			break;
		case 'g':
			rc = store_option (&cmdline, cmd_keyword_get_cache,
					   optarg);
			break;
		case 'c':
			rc = check_cache (optarg);
			if (rc >= 0) {
				rc = store_option (&cmdline, cmd_keyword_cache,
						   optarg);
			}
			break;
		case 'n':
			rc = check_no_cyl (optarg);
			if (rc >= 0) {
				rc = store_option (&cmdline, 
						   cmd_keyword_no_cyl,
						   optarg);
			}
			break;
		case 'p':
			rc = store_option(&cmdline, cmd_keyword_path, optarg);
			break;
		case OPT_PATH_RESET_ALL:
			rc = store_option(&cmdline, cmd_keyword_path_all,
					  optarg);
			break;
		case 'S':
			rc = store_option (&cmdline, cmd_keyword_reserve,
					   optarg);
			break;
		case 'L':
			rc = store_option (&cmdline, cmd_keyword_release,
					   optarg);
			break;
		case 'O':
			rc = store_option (&cmdline, cmd_keyword_slock,
					   optarg);
			break;
		case OPT_ENABLE_STATS:
			rc = store_option(&cmdline, cmd_keyword_enable_stats,
					  optarg);
			break;
		case OPT_DISABLE_STATS:
			rc = store_option(&cmdline, cmd_keyword_disable_stats,
					  optarg);
			break;
		case 'P':
			rc = store_option (&cmdline, cmd_keyword_profile,
					   optarg);
			break;
		case 'I':
			rc = check_prof_item (optarg);
			if (rc >= 0) {
				rc = store_option (&cmdline, 
						   cmd_keyword_prof_item,
						   optarg);
			}
			break;
		case 'R':
			rc = store_option (&cmdline, cmd_keyword_reset_prof,
					   optarg);
			break;
		case 'Q':
			rc = store_option (&cmdline, cmd_keyword_query_reserve,
					   optarg);
			break;
		case 's':
			rc = store_option(&cmdline, cmd_keyword_copy_swap, optarg);
			break;
		case -1:
			/* End of options string - start of devices list */
			cmdline.device_id = optind;
			break;
		default:
			fprintf(stderr, "Try 'tunedasd --help' for more"
					" information.\n");
			rc = -1;
			break;
		}
		if (rc < 0) {
			return rc;
		}
	} while (opt != -1);

	*line = cmdline;
	return 0;
}

/*
 * Execute the command.
 */
static int do_command(char *device, struct command_line cmdline)
{
	int i, rc;

	rc = 0;
	for (i = 0; !cmdline.parm[i].kw_given; i++);

	switch (i) {
	case cmd_keyword_get_cache:
                rc = disk_get_cache (device); 
		break;
	case cmd_keyword_cache:
		rc = disk_set_cache (device, 
				     cmdline.parm[cmd_keyword_cache].data,
				     cmdline.parm[cmd_keyword_no_cyl].data);
		break;
	case cmd_keyword_no_cyl:
		break;
	case cmd_keyword_reserve:
		rc = disk_reserve (device);
		break;
	case cmd_keyword_release:
		rc = disk_release (device);
		break;
	case cmd_keyword_slock:
		rc = disk_slock (device);
		break;
	case cmd_keyword_profile: 
		rc = disk_profile (device, 
				   cmdline.parm[cmd_keyword_prof_item].data);
		break;
	case cmd_keyword_reset_prof:
		rc = disk_reset_prof (device);
		break;
	case cmd_keyword_prof_item:
		break;
	case cmd_keyword_query_reserve:
		rc = disk_query_reserve_status(device);
		break;
	case cmd_keyword_path:
		rc = disk_reset_chpid(device,
				      cmdline.parm[cmd_keyword_path].data);
		break;
	case cmd_keyword_path_all:
		rc = disk_reset_chpid(device, NULL);
		break;
	case cmd_keyword_copy_swap:
		rc = disk_copy_swap(device, cmdline.parm[cmd_keyword_copy_swap].data);
		break;
	default:
		error_print ("Unknown command '%s' specified",
			     get_keyword_name (i));
		break;
	}

	return rc;
}

/*
 * Enable/Disable DASD performance statistics globally by writing
 * 'set on' or 'set off' to /proc/dasd/statistics.
 */
static int tunedasd_set_global_stats(int val)
{
	const char *path = "/proc/dasd/statistics";
	int rc = 0;

	if (val)
		rc = util_file_write_s("set on", path);
	else
		rc = util_file_write_s("set off", path);

	if (rc)
		error_print("Could not enable/disable performance statistics");
	else
		printf("Performance statistics %sabled\n", val ? "en" : "dis");

	return rc;
}

/*
 * Main. 
 */
int
main (int argc, char* argv[])
{
	struct command_line cmdline;
	int rc, finalrc;

	/* Find out what we're supposed to do */
	rc = get_command_line (argc, argv, &cmdline);
	if (rc) {
		return 1;
	}

	rc= check_key_state (&cmdline);
	if (rc) {
		return 1;
	}

	/* Check for priority options --help and --version */
	if (cmdline.parm[cmd_keyword_help].kw_given) {
		print_usage ();
		return 0;
	} else if (cmdline.parm[cmd_keyword_version].kw_given) {
		util_prg_print_version();
		return 0;
	}

	/* Make sure we're running as root */
	if (check_for_root ()) {
		return 1;
	}

	/* Enable/Disable performance statistics */
	if (cmdline.parm[cmd_keyword_enable_stats].kw_given)
		return tunedasd_set_global_stats(1);
	if (cmdline.parm[cmd_keyword_disable_stats].kw_given)
		return tunedasd_set_global_stats(0);

	/* Do each of the commands on each of the devices 
	 * and don't care about the return codes           */
	if (cmdline.device_id >= argc) {
		error_print ("Missing device");
		print_usage ();
		return 1;
	}

	finalrc = 0;
	while (cmdline.device_id < argc) {
		rc = do_command (argv[cmdline.device_id], cmdline);
		if (rc && !finalrc)
			finalrc = rc;
		cmdline.device_id++;
	}
	return finalrc;
}
